### DISCUSSION TITLE
Live Preview for Vue

### DISCUSSION THREAD

**Discussion by ChrisGV04** [payloadcms/payload#4970] [Feature Requests & Ideas]
I love the power of Live Preview. This feature alone has convinced some of my customers to switch from WordPress and other integrated CMS's. However, I work with Nuxt for Vue for my front-end development.

I see that the docs for Live Preview mention that you plan to officially support other frameworks, such as Vue.

Are you planning to implement this feature soon? Because I checked out the source code for the React hook and it seems straight forward to implement, so I would be willing to build it for Vue if you like.

**Comment by DanRibbens** [comment]
> I would be willing to build it for Vue if you like.

You are the first I know of offering to work on this. We would gladly accept your offer!

**Reply by ChrisGV04** [reply]
Sure thing! I'm glad to contribute. I'll start working on it later today

**Comment by ChrisGV04** [comment]
Hello @DanRibbens!
I created an initial implementation of the hook for Vue inside this repo: https://github.com/ChrisGV04/payload-live-preview-vue

I also published it to NPM as `@cgvweb/payload-live-preview-vue` so I could test it in several of my live projects and it worked perfectly on both dev and production builds.

However, I need a bit of guidance from you guys if you would like to make it an official Payload plugin... I didn't know how to setup the `package.json` to be consistent with your monorepo and how you do it for the other official plugins. That's why I built it in my own repo instead submitting a PR.

I have no experience with SWC nor monorepos, so I'm not sure if there's something I should setup differently. The hook is identical in nature to the official React hook, just with Vue syntax. It's also pure Typescript and doesn't require any Vue compilers.

**Reply by DanRibbens** [reply]
This is amazing!

Here is what we are going to do.
1. I'll create a branch that has our boilerplate plugin stuff added to the monorepo in `/packages/live-preview-vue`
2. You can branch from the new branch
3. Add your code to the new folder ^
4. Copy `/test/live-preview` to `test/live-preview-vue`
5. Copy whatever you need from an existing vue project
6. Commit and push your changes to your fork
7. Open the PR targeting `main`.

**Reply by ChrisGV04** [reply]
Thank you very much!
I'll start working on that.

**Reply by DanRibbens** [reply]
@ChrisGV04 

Here is the branch: https://github.com/payloadcms/payload/tree/feat/live-preview-vue

Let me know if you have questions or would prefer us doing this. I just want you to get the contribution credit! :)

**Reply by ChrisGV04** [reply]
Thanks! 

I'll give it a try and keep you posted.

**Reply by DanRibbens** [reply]
I just remembered, we can add this to the `/docs/live-preview` also.

**Reply by ChrisGV04** [reply]
Sure! I can add it as well. The README file of my repo was inspired by how the React hook is documented, so I can keep it with a similar format as the docs and use the example from my README since that's the exact equivalent to the React example.

I expect to submit the PR by Monday with tests and docs included.

**Reply by ChrisGV04** [reply]
Hi @DanRibbens!

Quick question about the tests…

I don’t see anything in the React tests that actually check if the data changes are displayed on the Live Preview iframe, it basically tests that `@payloadcms/live-preview` itself works, not the `@payloadcms/live-preview-react` hook.

So my question is: **Is it relevant to add a Vue test?**

You created a full Next app with pages, posts, blocks, rich texts, etc… Do I have to recreate the complete app on Vue? I don’t know if it is relevant since if it works on React, it works on Vue because there’s no Vue-specific logic happening here. At the end of the day `@payloadcms/live-preview` is framework agnostic, so the React test is equally valid for Vue.

**Reply by DanRibbens** [reply]
> I don’t see anything in the React tests that actually check if the data changes are displayed on the Live Preview iframe, it basically tests that @payloadcms/live-preview itself works, not the @payloadcms/live-preview-react hook.

Okay, I hadn't looked that closely at the existing e2e tests for live-preview and I wrongly assumed that it had this coverage and we would want the same. 

On the e2e side I see your point, the tests would be the same whether Vue or React clients. Could you open the PR with everything except the tests? I would like to see what it looks like and maybe we just need an integration testing on the functions instead of building up the vue app and testing that. Since we would have to also maintain the vue app in the test repo, we should avoid that unless needed for stability.

Thanks!

**Reply by ChrisGV04** [reply]
Sure! I will finish writing the docs for it and then submit the PR.

I know your main focus is on React, which is great. But if you ever need help maintaining Vue plugins or integrations for Payload (such as this hook), I'm happy to contribute.

**Reply by DanRibbens** [reply]
You are the best! I think a Vue frontend template could be very popular and if we were smart that could be used for e2e in the `test/live-preview-vue` setup.

**Reply by ChrisGV04** [reply]
Thank you for your kind words!

That sounds great! I can certainly try to recreate the [`website` starter template](https://github.com/payloadcms/payload/tree/main/templates/website) that you made with Next.js using Nuxt.

**Comment by ChrisGV04** [comment]
The PR has been submited #4987 

Please let me know if there's anything else missing. Thank you!

**Comment by DanRibbens** [comment]
I merged the PR and it will be released next week. Thank you @ChrisGV04 for this feature!

**Reply by ChrisGV04** [reply]
Glad to help!

**Comment by DanRibbens** [comment]
I cherry-picked your commit and updated the package to ESM for the PR against beta: https://github.com/payloadcms/payload/pull/5933 so that it can be included going forward.

We didn't add a test frontend for this. I wouldn't say it is required, but could be a nice to have. If you decide to add it you could skip v2 and just do it on v3, currently `beta`, will be the new `main` branch in a few weeks.

**Reply by ChrisGV04** [reply]
I agree on adding tests. I would love to see how you guys would test the React live preview hook so I can write a similar test for the Vue hook.

**Reply by DanRibbens** [reply]
Sure, we have a next frontend app in the live-preview test suite: https://github.com/payloadcms/payload/tree/main/test/live-preview. I would use the same config and replace all the Next app with Nuxt and replace the React components with Vue. It probably doesn't need to be nuxt if you just want to make it work as a vanilla Vue app that would be fine too. Alternatively you could just add unit tests to your component tests and avoid the work of making the full e2e.

Does that make sense?

**Reply by DanRibbens** [reply]
> I agree on adding tests. I would love to see how you guys would test the React live preview hook so I can write a similar test for the Vue hook.

@ChrisGV04 have you had a chance to look at this?

Even a simplified and barebones approach would be acceptable. As it stands we don't have a good way to release the package and ensure functionality in order to support it.

**Reply by ChrisGV04** [reply]
Hi @DanRibbens ! Sorry for the delay. I have been very busy lately.

I took a look at the current tests that exist for live preview and again I could not find any tests for the `live-preview-react` package specifically. The only test I could find was for the core `live-preview` and it never verifies that changes from the Admin UI get reflected on the `next-app`.

If I were to write a similar test, it would only ensure the Vue app runs, not that the `live-preview-vue` actually works.
